File: packages/sandbox/daemon/setup/autodetect.ts
import { existsSync } from "node:fs";
import { join } from "node:path";
import type { Application, PackageManager, RuntimeName } from "../types";

interface Detection {
  packageManager: PackageManager;
  runtime: RuntimeName;
}

const RULES: ReadonlyArray<{ file: string; detection: Detection }> = [
  { file: "deno.json", detection: { packageManager: "deno", runtime: "deno" } },
  {
    file: "deno.jsonc",
    detection: { packageManager: "deno", runtime: "deno" },
  },
  { file: "bun.lock", detection: { packageManager: "bun", runtime: "bun" } },
  { file: "bun.lockb", detection: { packageManager: "bun", runtime: "bun" } },
  {
    file: "pnpm-lock.yaml",
    detection: { packageManager: "pnpm", runtime: "node" },
  },
  {
    file: "yarn.lock",
    detection: { packageManager: "yarn", runtime: "node" },
  },
];

const NPM_FALLBACK: Detection = { packageManager: "npm", runtime: "node" };

/**
 * Best-effort runtime/pm guess from lockfile presence at the repo root.
 * Falls back to npm/node when no lockfile is recognised. Returns only the
 * fields that aren't already populated on the existing application config.
 */
export function autodetectApplication(
  repoDir: string,
  existing: Application | undefined,
): Partial<Application> {
  if (existing?.packageManager?.name && existing?.runtime) return {};

  const detected = detect(repoDir);
  return {
    ...(existing?.packageManager?.name
      ? {}
      : { packageManager: { name: detected.packageManager } }),
    ...(existing?.runtime ? {} : { runtime: detected.runtime }),
  };
}

function detect(repoDir: string): Detection {
  for (const rule of RULES) {
    if (existsSync(join(repoDir, rule.file))) return rule.detection;
  }
  return NPM_FALLBACK;
}

File: packages/sandbox/daemon-go/internal/setup/autodetect.go
package setup

import (
	"os"
	"path/filepath"
)

type Detection struct {
	PackageManager string
	Runtime        string
}

var detectRules = []struct {
	File      string
	Detection Detection
}{
	{"deno.json", Detection{"deno", "deno"}},
	{"deno.jsonc", Detection{"deno", "deno"}},
	{"bun.lock", Detection{"bun", "bun"}},
	{"bun.lockb", Detection{"bun", "bun"}},
	{"pnpm-lock.yaml", Detection{"pnpm", "node"}},
	{"yarn.lock", Detection{"yarn", "node"}},
}

var npmFallback = Detection{"npm", "node"}

func Autodetect(repoDir string) Detection {
	for _, rule := range detectRules {
		if _, err := os.Stat(filepath.Join(repoDir, rule.File)); err == nil {
			return rule.Detection
		}
	}
	return npmFallback
}
